### ISSUE TITLE
Phase 1 Stage 4: Data Exploration Framework - Comprehensive Implementation

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#22]
# Phase 1 Stage 4: Data Exploration Framework

## Overview
Implement comprehensive data exploration framework for steel casting defect prediction system. This notebook will serve as the foundation for understanding synthetic data patterns, validating data generation logic, and informing feature engineering decisions for both baseline XGBoost and LSTM models.

## File Location
`notebooks/01_data_exploration.ipynb`

### 1. Sensor Time Series Visualization
- **Normal vs Defect Cast Comparison**: Create side-by-side time series plots comparing sensor readings between good and defective casts
- **Multi-sensor Dashboard**: Implement interactive plots showing all sensor channels simultaneously
- **Pattern Recognition**: Identify visual patterns that distinguish defective casting sequences
- **Temporal Analysis**: Examine how sensor readings evolve throughout the casting process

### 2. Statistical Distribution Analysis
- **Sensor Value Distributions**: Generate histograms and box plots for each sensor channel
- **Defect Class Stratification**: Compare distributions between good/defect classes
- **Outlier Detection**: Identify anomalous sensor readings that may indicate defects
- **Statistical Tests**: Perform Kolmogorov-Smirnov tests to compare distributions

### 3. Correlation Matrices Between Sensors
- **Cross-Sensor Correlation Heatmaps**: Visualize correlations between all sensor pairs
- **Defect-Specific Correlations**: Compare correlation patterns in good vs defective casts
- **Time-Lagged Correlations**: Analyze delayed relationships between sensors
- **Feature Importance Indicators**: Identify which sensor combinations are most predictive

### 4. Defect Labeling Validation
- **Label Distribution Analysis**: Examine balance between good/defect classes
- **Labeling Logic Verification**: Validate that synthetic defect conditions align with domain knowledge
- **Edge Case Analysis**: Identify borderline cases that may be mislabeled
- **Domain Expert Review**: Document findings for validation against steel casting expertise

### 5. Data Quality Assessment
- **Missing Value Analysis**: Check for gaps in sensor data
- **Data Consistency Checks**: Verify sensor readings are within expected ranges
- **Temporal Continuity**: Ensure proper time sequencing in generated data
- **Synthetic Data Realism**: Compare generated patterns to expected steel casting behavior

### Required Libraries
```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from scipy import stats
from sklearn.preprocessing import StandardScaler
```

### Key Visualizations
1. **Multi-panel Time Series**: 
   - Temperature profiles (mold, secondary cooling zones)
   - Mechanical parameters (casting speed, oscillation frequency)
   - Process conditions (pressure, flow rates)

2. **Statistical Comparison Plots**:
   - Distribution overlays for good vs defect casts
   - Correlation heatmaps with hierarchical clustering
   - Box plots showing quartile distributions by defect class

3. **Interactive Dashboards**:
   - Plotly-based interactive sensor exploration
   - Filtering capabilities by cast_id, time ranges, defect status
   - Zoom and pan functionality for detailed inspection

### Domain-Specific Analysis

Based on the steel casting process understanding:

#### Thermal Conditions Analysis
- **Superheat Temperature**: Analyze temperature above liquidus point
- **Mold Heat Flux**: Examine heat transfer patterns in mold region
- **Secondary Cooling**: Evaluate spray cooling effectiveness
- **Cooling Rate Calculations**: Derive cooling rates from temperature profiles

#### Mechanical/Operational Parameters
- **Casting Speed Stability**: Analyze speed variations and their impact
- **Mold Level Control**: Examine mold level fluctuations
- **Oscillation Parameters**: Study frequency and stroke patterns
- **Withdrawal Forces**: Analyze mechanical stress indicators

#### Quality Defect Correlation
Map sensor patterns to specific defect types:
- **Surface Cracks**: Correlate with thermal gradients and mechanical stress
- **Internal Cracks**: Link to centerline cooling patterns
- **Corner Cracks**: Associate with mold corner heat transfer
- **Shape Deformities**: Connect to withdrawal force irregularities
- **Inclusion Entrapment**: Relate to flow pattern disruptions

### 1. Data Summary Report
- Comprehensive statistics for all sensor channels
- Defect rate analysis and class distribution
- Data quality metrics and validation results

### 2. Visualization Portfolio
- Publication-quality plots for technical documentation
- Interactive dashboards for stakeholder presentations
- Comparative analysis charts for model development guidance

### 3. Feature Engineering Recommendations
- Identification of most informative sensor combinations
- Suggested time window sizes for feature extraction
- Statistical feature candidates (mean, std, min, max, median)
- Stability indicators (spike counts, excursion frequencies)

### 4. Model Development Insights
- Baseline model feature prioritization
- LSTM sequence length recommendations
- Data preprocessing requirements
- Class imbalance handling strategies

## Success Criteria

- [ ] Complete exploratory analysis of all synthetic sensor channels
- [ ] Validation of defect labeling logic against domain knowledge
- [ ] Identification of key sensor patterns distinguishing good/defect casts
- [ ] Documentation of data quality assessment results
- [ ] Generation of actionable insights for Phase 2 feature engineering
- [ ] Creation of reusable visualization templates for ongoing analysis

## Dependencies
- Synthetic data generation must be completed (Phase 1, Stage 1-3)
- Access to steel casting domain expertise for validation
- Jupyter notebook environment with required visualization libraries

## Timeline
This analysis should be completed within Phase 1 (Week 2) as specified in the project milestones, serving as the foundation for subsequent feature engineering and model development phases.

## Integration with Project Architecture
This notebook will integrate with:
- `src/data/data_generator.py` for loading synthetic datasets
- `src/features/feature_engineer.py` for informing feature extraction logic
- `configs/data_generation.yaml` for understanding data generation parameters
- Future model development notebooks for baseline and LSTM implementations
